### ISSUE TITLE
OpenSSL 3.0.0alpha16 interop issue with 0.9.8zh - Diffie-Hellman invalid public key

### ISSUE THREAD

**Issue by rainerjung** [openssl/openssl#15465]
I ran into an interop problem when doing release testing for httpd 2.4.48 using OpenSSL 3.0.0alpha16 in the server and OpenSSL 0.9.8zh in the client. First I made TLS 1.0 work by adding "@SECLEVEL=0" at the end of the SSLCipherSuite config lines, but I still see sporadic failures:

SSL Library Error: error:02800066:Diffie-Hellman routines::invalid public key

I think they might have to do with restricting the allowed params for DH in OpenSSL 3. So it seems sometimes 0.9.8 generates DH keys that do not fulfil the stricter requirements of version 3. That means using such old client software will most of the time succeed with an TLS 1.0 handshake, but sporadically it will fail. I am not arware of a workaround.

Of course TLS 1.0 should no longer be used, but since it is supported in OpenSSL 3, the behavior of sporadic interop failures is a bit problematic and should at least be documented.

With a bit of help how to add more debug info, I could do a custom httpd build. But I will try to reproduce with s_server and s_client first and hope there's enough debug flags there to understand the used DH params.

**Comment by mattcaswell** [comment]
> ossl_cms_ctx_get0_propq(ctx)

Thanks @rainerjung - that would be very helpful

**Comment by rainerjung** [comment]
I can not reproduce with 0.9.8 s_client, but with a very simple programmatic client using 0.9.8.

I am convinced, that the problem is due to the missing padding (DH_compute_key_padded(()) in OpenSSL before 1.0.2 introduced in this commit first in 1.0.2:

https://github.com/openssl/openssl/commit/6c4b3514d740020be9bf7c757930f8ca536622ee

OpenSSL 3 sets DH_R_INVALID_PUBKEY in crypto/dh/dh_key.c line 392 when the size of the public key is not correct. Although the error only happens spuriously it happens roughly 1/256 of the attempts, which would be a good fit to the average likelyness, that the key has at least one byte less as size.

The stricter check in 3 was added in https://github.com/openssl/openssl/commit/9aaecbfc98eb89a03f72b35d343e08f377e7803a as a side effect of FFDHE support. Since the padding is mentioned in section 4.2.8.1 of RFC 8446 and was introduced in 1.0.2 as a client in 2013 the interop break with older OpenSSL might be OK.

A very verbose dump of s_server shows the following differences between a working handshake and a breaking one:

Working:

SSL_accept:SSLv3/TLS write server done
read from 0x1570270 [0x14e5858] (262 bytes => 262 (0x106))
...
<<< TLS 1.0, Handshake [length 0106], ClientKeyExchange
...
read from 0x1570270 [0x14e5853] (5 bytes => 5 (0x5))
0000 - 14 03 01 00 01                                    .....
<<< TLS 1.0, RecordHeader [length 0005]
    14 03 01 00 01
SSL_accept:SSLv3/TLS read client key exchange
read from 0x1570270 [0x14e5858] (1 bytes => 1 (0x1))
0000 - 01                                                .
read from 0x1570270 [0x14e5853] (5 bytes => 5 (0x5))
0000 - 16 03 01 00 30                                    ....0
<<< TLS 1.0, RecordHeader [length 0005]
    16 03 01 00 30
SSL_accept:SSLv3/TLS read change cipher spec

Broken:

SSL_accept:SSLv3/TLS write server done
read from 0x154d590 [0x1505898] (261 bytes => 261 (0x105))
...
<<< TLS 1.0, Handshake [length 0105], ClientKeyExchange
...
>>> TLS 1.0, RecordHeader [length 0005]
    15 03 01 00 02
write to 0x154d590 [0x14ebdc0] (7 bytes => 7 (0x7))
0000 - 15 03 01 00 02 02 50                              ......P
>>> TLS 1.0, Alert [length 0002], fatal internal_error
    02 50
SSL3 alert write:fatal:internal error
SSL_accept:error in error
00879386C87F0000:error:02800066:Diffie-Hellman routines:(unknown function):invalid public key:crypto/dh/dh_key.c:399:
00879386C87F0000:error:0A0C0103:SSL routines:(unknown function):internal error:ssl/statem/statem_srvr.c:2983:

**Comment by arekm** [comment]
This works as reproducer for me. Server side with openssl 3.0.1:
````
openssl s_server --accept *:5666 --nocert -cipher "ALL:@SECLEVEL=0"
````

client side openssl 1.1.1l:
````
while (true); do echo "x" | openssl s_client --connect server:5666 -cipher "ALL:@SECLEVEL=0" || break; done
````

Once a while server reports:
````
80BB0D5D387F0000:error:02800066:Diffie-Hellman routines:ossl_dh_buf2key:invalid public key:crypto/dh/dh_key.c:399:
80BB0D5D387F0000:error:0A0C0103:SSL routines:tls_process_cke_dhe:internal error:ssl/statem/statem_srvr.c:2989:
`````

Not seeing this between 3.0.1 <--> 3.0.1 and 1.1.1l <--> 1.1.1l. Only when 3.0.1 is server and 1.1.1l is a client (not vice versa)

**Comment by t8m** [comment]
Fix in #17630
